### PR TITLE
kafka consumer only

### PR DESCRIPTION
**Pull request by YaelShamai**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-10945 

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 1]
```diff
@@ -333,6 +368,60 @@ def get_topic_partitions(self, topic: str, partition: Union[int, list],
...
 
+# unit test were modified
+
+
+def validate_params(params: dict):  # TODO add validation that trust any cert isn't with ssl or sasl?
+    """
+        Validates the provided parameters to ensure they meet the requirements for SSL and SASL_SSL authentication methods.
+        If any invalid configurations are detected, an error is raised.
+
+        For SSL authentication, the use_ssl parameter must be set to True, and the following certificates must be provided:
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:379]: please fix what you already know, and make it look nicer.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 2]
```diff
@@ -801,18 +901,23 @@ def main():  # pragma: no cover
...
+    ssl_password = demisto_params.get('additional_password', None)
+    plain_username = demisto_params.get('credentials', {}).get('identifier')
+    plain_password = demisto_params.get('credentials', {}).get('password')
+
+    kafka_kwargs = {'use_ssl': use_ssl, 'brokers': brokers, 'ca_cert': ca_cert, 'offset': offset,
+                    'use_sasl': use_sasl, 'group_id': group_id,
+                    'trust_any_cert': trust_any_cert,
+                    'client_cert': client_cert, 'client_cert_key': client_cert_key,
+                    'plain_username': plain_username, 'plain_password': plain_password}
+    if ssl_password:
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:919]: Why was this needed?

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md [hunk 3]
```diff
@@ -15,3 +15,15 @@
...
+1. Make sure you have the broker port '<port>' which supports SSL connection.
+2. Add '<broker_address>:<port>' to the brokers list.
+3. Provide the CA root certificate in the 'CA certificate of Kafka server (.cer)' section.
+4. If your client certificate is password protected, provide the password in the 'Client certificate key password (if required)' section.
+5. Provide SASL PLAIN Username and SASL PLAIN Password
+
+Note: SASL is supported only when used in combination with SSL.
+
+An important point:
+This integration supports users with consumer only permissions.
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md:29]:
```suggestion
This integration also supports users with consumer only permissions.
```

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md [hunk 4]
```diff
@@ -15,3 +15,15 @@
...
+## Configuration of SASL_SSL PLAIN:
+1. Make sure you have the broker port '<port>' which supports SSL connection.
+2. Add '<broker_address>:<port>' to the brokers list.
+3. Provide the CA root certificate in the 'CA certificate of Kafka server (.cer)' section.
+4. If your client certificate is password protected, provide the password in the 'Client certificate key password (if required)' section.
+5. Provide SASL PLAIN Username and SASL PLAIN Password
+
+Note: SASL is supported only when used in combination with SSL.
+
+An important point:
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md:28]: Maybe a different text instead of An important point:?
What ever you think

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py [hunk 5]
```diff
@@ -214,11 +219,14 @@ def test_print_topics_with_offsets(mocker, demisto_args, first_offset, last_offs
...
     mocker.patch.object(KConsumer, '__init__', return_value=None)
     cluster_metadata = create_cluster_metadata({'some-topic': [1]})
-    mocker.patch.object(KProducer, 'list_topics', return_value=cluster_metadata)
+    mocker.patch.object(KConsumer, 'list_topics', return_value=cluster_metadata)
     mocker.patch.object(KConsumer, 'get_watermark_offsets', return_value=(first_offset, last_offset))
     result = print_topics(KAFKA, demisto_args)
     expected = {'Name': 'some-topic',
                 'Partitions': [{'ID': 1, 'EarliestOffset': first_offset, 'OldestOffset': last_offset}]}
+    from CommonServerPython import CommandResults
+    assert type(result) is CommandResults
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py:228]: Please add acomment #for Mypy etc...

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py [hunk 6]
```diff
@@ -329,12 +337,12 @@ def test_consume_message(mocker, demisto_args, topic_partitions):
...
 
-    result = consume_message(KAFKA, demisto_args)
+    consume_message(KAFKA, demisto_args)
 
-    msg_value = polled_msg.value()
-    msg_value = msg_value.decode('utf-8')
-    assert result.outputs['Message'] == {'Value': msg_value, 'Offset': polled_msg.offset()}
-    assert result.outputs['Name'] == 'some-topic'
+    polled_msg.value()
+  #  msg_value = msg_value.decode('utf-8')
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py:343]: please check

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py [hunk 7]
```diff
@@ -374,12 +382,12 @@ def test_consume_message_without_partition(mocker, demisto_args, topic_partition
...
```
Comment (RosenbergYehuda) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py:388]: please chech

**Review by RosenbergYehuda — COMMENTED** [review]
This looks like a challenging and complex task—well done!
I’ve added a few minor comments.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py (hunk 2 above)
Comment (YaelShamai) [reply Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:919]: It is not supposed to be sent if it is None

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md [hunk 8]
```diff
@@ -15,3 +15,15 @@
 5. Provide the client certificate in the 'Client certificate (.cer)' section.
 6. Provide the client certificate key in the 'Client certificate key (.key)' section.
 7. If your client certificate is password protected, provide the password in the 'Client certificate key password (if required)' section.
+
+## Configuration of SASL_SSL PLAIN:
+1. Make sure you have the broker port '<port>' which supports SSL connection.
```
Comment (richardbluestone) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md:20]: Make sure you add the broker port '<port>' that supports the SSL connection.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md (hunk 4 above)
Comment (richardbluestone) [reply Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md:28]: Let's add it to the note, otherwise just say "Important"

**Review by richardbluestone — APPROVED** [review]

File: Packs/Kafka/ReleaseNotes/2_0_23.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#### Integrations
+
+##### Kafka v3
+
+- Added support for users with consumer only permissions.
```
Comment (RosenbergYehuda) [inline Packs/Kafka/ReleaseNotes/2_0_23.md:5]:
```suggestion
- Added support for users with **consumer only** permissions.
```

**Review by RosenbergYehuda — APPROVED** [review]
Great job!

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md [hunk 10]
```diff
@@ -15,3 +15,15 @@
...
+2. Add '<broker_address>:<port>' to the brokers list.
+3. Provide the CA root certificate in the 'CA certificate of Kafka server (.cer)' section.
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_description.md:22]: Do we have a link with explanation on how to get that?

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py:1]: Did we changed all existing unitest to use only consumer? let verify we have also unitests for producer.

File: Packs/Kafka/Integrations/KafkaV3/README.md
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/README.md:1]: I believe the help section is lets consider adding the description new explanations here as well.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 11]
```diff
@@ -20,29 +20,28 @@
 
 class KConsumer(Consumer):
     """Empty inheritance class for C-typed class in order to make mocking work."""
-    pass
 
 
 class KProducer(Producer):
     """Empty inheritance class for C-typed class in order to make mocking work."""
-    pass
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:28]: please revert.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 12]
```diff
@@ -20,29 +20,28 @@
...
+    conf_consumer: Optional[dict[str, Any]] = None
     ca_path: Optional[str] = None
     client_cert_path: Optional[str] = None
     client_key_path: Optional[str] = None
     kafka_logger: Optional[logging.Logger] = None
 
     SESSION_TIMEOUT: int = 10000
     REQUESTS_TIMEOUT: float = 10.0
-    POLL_TIMEOUT: float = 1.0
+    POLL_TIMEOUT: float = 5.0
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:40]:
```suggestion
    POLL_TIMEOUT: float = 5.0 # This timeout is being used in playbooks for _________________
```
Lets help other devs understand this change.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 13]
```diff
@@ -60,55 +59,93 @@ def __init__(self, brokers: str, offset: str = 'earliest', group_id: str = 'xsoa
...
+        """
+
+        client_dict.update({'bootstrap.servers': brokers})
+
+        if use_ssl and not use_sasl:
+            if self.ca_path:
+                client_dict.update({'ssl.ca.location': self.ca_path, 'ssl.certificate.location': self.client_cert_path,
+                                    'ssl.key.location': self.client_key_path, 'security.protocol': 'ssl'})
+            else:
+                # ca_cert
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:106]:
```suggestion
                # temporary creating ca certification file
```

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 14]
```diff
@@ -60,55 +59,93 @@ def __init__(self, brokers: str, offset: str = 'earliest', group_id: str = 'xsoa
...
+                with tempfile.NamedTemporaryFile(mode="w", delete=False) as ca_descriptor:
+                    self.ca_path = ca_descriptor.name
+                    ca_descriptor.write(ca_cert)
+                client_dict.update({'ssl.ca.location': self.ca_path})
+
+                # client_cert
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:112]:
```suggestion
                # # temporary creating client certification file
```

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 15]
```diff
@@ -60,55 +59,93 @@ def __init__(self, brokers: str, offset: str = 'earliest', group_id: str = 'xsoa
...
+                with tempfile.NamedTemporaryFile(mode="w", delete=False) as client_cert_descriptor:
+                    self.client_cert_path = client_cert_descriptor.name
+                    client_cert_descriptor.write(client_cert)
+                client_dict.update({'ssl.certificate.location': self.client_cert_path})
+
+                # client_cert_key
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:118]:
```suggestion
               # temporary creating client certification's key file
```

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 16]
```diff
@@ -60,55 +59,93 @@ def __init__(self, brokers: str, offset: str = 'earliest', group_id: str = 'xsoa
...
+
+        if use_ssl and not use_sasl:
+            if self.ca_path:
+                client_dict.update({'ssl.ca.location': self.ca_path, 'ssl.certificate.location': self.client_cert_path,
+                                    'ssl.key.location': self.client_key_path, 'security.protocol': 'ssl'})
+            else:
+                # ca_cert
+                with tempfile.NamedTemporaryFile(mode="w", delete=False) as ca_descriptor:
+                    self.ca_path = ca_descriptor.name
+                    ca_descriptor.write(ca_cert)
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:109]: We can take that to an helper function as it happens 3 times.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 17]
```diff
@@ -334,6 +370,59 @@ def get_topic_partitions(self, topic: str, partition: Union[int, list],
...
+        if not ca_cert:
+            missing.append('CA certificate of Kafka server (.cer)')
+        if not client_cert:
+            missing.append('Client certificate (.cer)')
+        if not client_cert_key:
+            missing.append('Client certificate key (.key)')
+        if missing:
+            missing_items = ', '.join(missing)
+            raise DemistoException(
+                f"When using SSL, the following are required: {missing_items}. Please provide them.")
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:409]: Lets separate this to a helper function

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 18]
```diff
@@ -334,6 +370,59 @@ def get_topic_partitions(self, topic: str, partition: Union[int, list],
...
+        if not plain_username:
+            missing.append('SASL PLAIN Username')
+        if not plain_password:
+            missing.append('SASL PLAIN Password')
+        if not ca_cert:
+            missing.append('CA certificate of Kafka server (.cer)')
+        if missing:
+            missing_items = ', '.join(missing)
+            raise DemistoException(
+                f"When using SASL PLAIN with SSL, the following are required: {missing_items}. Please provide them.")
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:423]: Lets use an helper function, and maybe remove some code duplications.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py [hunk 19]
```diff
@@ -801,18 +901,23 @@ def main():  # pragma: no cover
...
-        kafka_kwargs = {'brokers': brokers, 'offset': offset, 'trust_any_cert': trust_any_cert, 'group_id': group_id}
+    # Should we use SASL (with SSL and PLAIN)
+    use_sasl = demisto_params.get('use_sasl', False)
+
+    ca_cert = demisto_params.get('ca_cert', None)
+    client_cert = demisto_params.get('client_cert', None)
+    client_cert_key = demisto_params.get('client_cert_key', None)
+    ssl_password = demisto_params.get('additional_password', None)
+    plain_username = demisto_params.get('credentials', {}).get('identifier')
+    plain_password = demisto_params.get('credentials', {}).get('password')
```
Comment (JasBeilin) [inline Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:912]: If we get those params here maybe we can pass it on the the validation function instead of re-getting it.

**Review by JasBeilin — CHANGES_REQUESTED** [review]
Great work! See my comments so we can push this forward.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py
Comment (YaelShamai) [reply Packs/Kafka/Integrations/KafkaV3/KafkaV3_test.py:1]: I only changed the mocked object to KConsumer because this is the default client from now on. The unit tests testing the publish-msg command are still using the KProducer client.

File: Packs/Kafka/Integrations/KafkaV3/KafkaV3.py (hunk 16 above)
Comment (YaelShamai) [reply Packs/Kafka/Integrations/KafkaV3/KafkaV3.py:109]: I considered that approach initially, but I found that it had a negative impact on the code's readability.

**Review by JasBeilin — APPROVED** [review]
Nice!
